File: 1.2-ans/regression_1-2.py
import sys, math
import numpy as np
import pystan
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

def simulate():
  print '\tGenerating simulated data...'

  N = 30
  D = 10

  weights = [x - (D/2.0) for x in range(D)]
  weights = np.array(weights)
  print 'True Weights:', weights

  fmean, fstd = 0, 1
  features = np.zeros((N, D))
  for i in range(N):
    for j in range(D):
      features[i][j] = fstd * np.random.randn() + fmean

  observations = np.dot(features, weights)

  sigma = 1
  noise = np.zeros((observations.shape))
  for i in range(len(noise)):
    noise[i] = sigma * np.random.randn() + 0

  observations += noise

  # Construct dictionary to pass to Stan
  dataset = {
    'N': N,
    'D': D,
    'observations': observations,
    'features': features
  }
  return dataset


def inference(dataset):
  print '\tPerforming inference...'
  NUM_ITER = 2000
  WARMUP = 200
  NUM_CHAINS = 4
  NUM_CORES = 4
  STAN_FN = 'regression_1-2.stan'

  # import pdb; pdb.set_trace()
  fit = pystan.stan(file = STAN_FN, 
                    data = dataset, 
                    iter = NUM_ITER, 
                    warmup = WARMUP, 
                    chains = NUM_CHAINS, 
                    n_jobs = NUM_CORES)
  print(fit)

  fit.plot()
  plt.tight_layout()
  plt.savefig('fit.png')
  return


def main():
  dataset = simulate()
  inference(dataset)
  return


if __name__ == '__main__':
  main()

File: 1.2-ans/regression_1-2.R
library('rstan');

N <- 30;
D <- 10;
weights <- c(-5, -4, -3, -2, -1, -0, 1, 2, 3, 4);
normal_sigma <- 1;

feature_mean <- 0;
feature_sigma <- 1;

#########################################################
# Simulate data
#########################################################

features <- matrix(, nrow = N, ncol = D);
for (row in 1:N) {
  features[row, ] <- rnorm(D, 
                           mean = feature_mean, 
                           sd = feature_sigma);
}

obs <- features %*% weights;    # matrix multiply
noise <- rnorm(N, mean = 0, sd = normal_sigma);
obs <- obs + noise;
obs <- as.vector(obs);   # convert to vector

#########################################################
# Run inference
#########################################################

dataset <- list(N = N, 
                D = D,
                observations = obs,
                features = features);

fit <- stan(file = 'regression_1-2.stan', 
            data = dataset, 
            iter = 1000, 
            chains = 4);
print(fit);

png(filename = 'fit_rstan.png')
plot(fit);
dev.off();
